### ISSUE TITLE
Error building on Ubuntu 19.04

### ISSUE THREAD

**Issue by bsespede** [weigao95/surfelwarp#26]
### **System**

- Ubuntu 19.04
- Using eigen from libeigen3-dev. Tried building it myself but I get the same error
- CUDA 10.1
- Latest commits from `cilantro` and `Pangolin` as of today.

### **Error**

During compilation I get these errors when building `common_lib_generated_sanity_check.cu.o`

### **Output**

[  1%] Building NVCC (Device) object common/CMakeFiles/common_lib.dir/common_lib_generated_sanity_check.cu.o:
_In file included from /usr/include/eigen3/Eigen/Core:257:0,
                 from /usr/include/eigen3/Eigen/Dense:1,
                 from /usr/include/eigen3/Eigen/Eigen:1,
                 from /home/bsespede/Repositories/surfelwarp/common/common_types.h:16,
                 from /home/bsespede/Repositories/surfelwarp/common/sanity_check.cu:1:
/usr/include/host_defines.h:54:2: warning: #warning "host_defines.h is an internal header file and must not be used directly.  This file will be removed in a future CUDA release.  Please use cuda_runtime_api.h or cuda_runtime.h instead." [-Wcpp]
 #warning "host_defines.h is an internal header file and must not be used directly.  This file will be removed in a future CUDA release.  Please use cuda_runtime_api.h or cuda_runtime.h instead."
  ^~~~~~~
In file included from /usr/include/eigen3/Eigen/Core:257:0,
                 from /usr/include/eigen3/Eigen/Dense:1,
                 from /usr/include/eigen3/Eigen/Eigen:1,
                 from /home/bsespede/Repositories/surfelwarp/common/common_types.h:16,
                 from /home/bsespede/Repositories/surfelwarp/common/sanity_check.cu:1:
/usr/include/host_defines.h:54:2: warning: #warning "host_defines.h is an internal header file and must not be used directly.  This file will be removed in a future CUDA release.  Please use cuda_runtime_api.h or cuda_runtime.h instead." [-Wcpp]
 #warning "host_defines.h is an internal header file and must not be used directly.  This file will be removed in a future CUDA release.  Please use cuda_runtime_api.h or cuda_runtime.h instead."
  ^~~~~~~
In file included from /usr/include/eigen3/Eigen/Core:257:0,
                 from /usr/include/eigen3/Eigen/Dense:1,
                 from /usr/include/eigen3/Eigen/Eigen:1,
                 from /home/bsespede/Repositories/surfelwarp/common/common_types.h:16,
                 from /home/bsespede/Repositories/surfelwarp/common/sanity_check.cu:1:
/usr/include/host_defines.h:54:2: warning: #warning "host_defines.h is an internal header file and must not be used directly.  This file will be removed in a future CUDA release.  Please use cuda_runtime_api.h or cuda_runtime.h instead." [-Wcpp]
 #warning "host_defines.h is an internal header file and must not be used directly.  This file will be removed in a future CUDA release.  Please use cuda_runtime_api.h or cuda_runtime.h instead."
  ^~~~~~~
In file included from /usr/include/eigen3/Eigen/Core:257:0,
                 from /usr/include/eigen3/Eigen/Dense:1,
                 from /usr/include/eigen3/Eigen/Eigen:1,
                 from /home/bsespede/Repositories/surfelwarp/common/common_types.h:16,
                 from /home/bsespede/Repositories/surfelwarp/common/sanity_check.cu:1:
/usr/include/host_defines.h:54:2: warning: #warning "host_defines.h is an internal header file and must not be used directly.  This file will be removed in a future CUDA release.  Please use cuda_runtime_api.h or cuda_runtime.h instead." [-Wcpp]
 #warning "host_defines.h is an internal header file and must not be used directly.  This file will be removed in a future CUDA release.  Please use cuda_runtime_api.h or cuda_runtime.h instead."
  ^~~~~~~
In file included from /usr/include/eigen3/Eigen/Core:257:0,
                 from /usr/include/eigen3/Eigen/Dense:1,
                 from /usr/include/eigen3/Eigen/Eigen:1,
                 from /home/bsespede/Repositories/surfelwarp/common/common_types.h:16,
                 from /home/bsespede/Repositories/surfelwarp/common/sanity_check.cu:1:
/usr/include/host_defines.h:54:2: warning: #warning "host_defines.h is an internal header file and must not be used directly.  This file will be removed in a future CUDA release.  Please use cuda_runtime_api.h or cuda_runtime.h instead." [-Wcpp]
 #warning "host_defines.h is an internal header file and must not be used directly.  This file will be removed in a future CUDA release.  Please use cuda_runtime_api.h or cuda_runtime.h instead."
  ^~~~~~~
/usr/include/eigen3/Eigen/src/Core/MathFunctions.h(1075): warning: calling a constexpr __host__ function("real") from a __host__ __device__ function("abs") is not allowed. The experimental flag '--expt-relaxed-constexpr' can be used to allow this.

/usr/include/eigen3/Eigen/src/Core/MathFunctions.h(1075): warning: calling a constexpr __host__ function("imag") from a __host__ __device__ function("abs") is not allowed. The experimental flag '--expt-relaxed-constexpr' can be used to allow this.

/usr/include/eigen3/Eigen/src/Core/MathFunctions.h(1075): warning: calling a constexpr __host__ function from a __host__ __device__ function is not allowed. The experimental flag '--expt-relaxed-constexpr' can be used to allow this.

/usr/include/eigen3/Eigen/src/Core/MathFunctions.h(1075): warning: calling a constexpr __host__ function from a __host__ __device__ function is not allowed. The experimental flag '--expt-relaxed-constexpr' can be used to allow this.

/usr/include/eigen3/Eigen/src/Core/MathFunctions.h(1080): warning: calling a constexpr __host__ function("real") from a __host__ __device__ function("abs") is not allowed. The experimental flag '--expt-relaxed-constexpr' can be used to allow this.

/usr/include/eigen3/Eigen/src/Core/MathFunctions.h(1080): warning: calling a constexpr __host__ function("imag") from a __host__ __device__ function("abs") is not allowed. The experimental flag '--expt-relaxed-constexpr' can be used to allow this.

/usr/include/eigen3/Eigen/src/Core/MathFunctions.h(1080): warning: calling a constexpr __host__ function from a __host__ __device__ function is not allowed. The experimental flag '--expt-relaxed-constexpr' can be used to allow this.

/usr/include/eigen3/Eigen/src/Core/MathFunctions.h(1080): warning: calling a constexpr __host__ function from a __host__ __device__ function is not allowed. The experimental flag '--expt-relaxed-constexpr' can be used to allow this.

/usr/include/eigen3/Eigen/src/Core/arch/CUDA/Half.h(302): warning: calling a __host__ function from a __host__ __device__ function is not allowed

/usr/include/eigen3/Eigen/src/Core/arch/CUDA/Half.h(305): warning: calling a __host__ function from a __host__ __device__ function is not allowed

/usr/include/eigen3/Eigen/src/Core/arch/CUDA/PacketMathHalf.h(81): error: no instance of overloaded function "__ldg" matches the argument list
            argument types are: (const Eigen::half *)

/usr/include/eigen3/Eigen/src/Core/arch/CUDA/PacketMathHalf.h(81): error: no instance of overloaded function "__ldg" matches the argument list
            argument types are: (const Eigen::half *)

2 errors detected in the compilation of "/tmp/tmpxft_00001cb4_00000000-8_sanity_check.compute_61.cpp1.ii".
CMake Error at common_lib_generated_sanity_check.cu.o.cmake:279 (message):
  Error generating file
  /home/bsespede/Repositories/surfelwarp/build/common/CMakeFiles/common_lib.dir//./common_lib_generated_sanity_check.cu.o

make[2]: *** [common/CMakeFiles/common_lib.dir/build.make:2516: common/CMakeFiles/common_lib.dir/common_lib_generated_sanity_check.cu.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:108: common/CMakeFiles/common_lib.dir/all] Error 2
make: *** [Makefile:84: all] Error 2_

**Comment by weigao95** [comment]
That seems to be the problem of using Eigen in CUDA. Can you try to write a simple .cu file that include Eigen and make sure it works?

Sorry but I don't have Ubuntu 19 environment available.

**Comment by bsespede** [comment]
It was Indeed an incompatibility issue between CUDA 10.2 and Eigen 3.7. Had to downgrade to CUDA 9.2.

Now I'm having issues with during linking with cilantro, as apparently the file <cilantro/point_cloud.hpp> was moved in recent versions. May I ask you which commit you used for cilantro? Otherwise I'll give PCL a go.

Thanks for your time!

**Comment by bsespede** [comment]
Hi Wei,

I've finally managed to find a working cilantro version and it builds perfectly fine. However, after I downloaded and tried to run the sample 'boxing' scene I get the following error during runtime:

`/home/bsespede/Repositories/surfelwarp/cmake-build-debug/apps/surfelwarp_app/surfelwarp_app ./data/boxing_config.json
Error: 	/home/bsespede/Repositories/surfelwarp/common/algorithm_types.h:85`

After checking that file, it seems to crash during the construction of `FlagSelection`, so it seems CUDA related? I'm using CUDA 9.2 for reference.
